app/waf/src/controllers/application.controller.ts

```diff
@@ -155,4 +184,115 @@ export class ApplicationController {
   async deleteById(@param.path.string('id') id: string): Promise<void> {
     await this.applicationRepository.deleteById(id);
   }
+
+  @post(prefix + '/applications/{id}/deploy', {
+    responses: {
+      '204': {
+        description: 'Application deploy success',
+      },
+    },
+  })
+  async deployById(@param.path.string('id') id: string): Promise<Object> {
+    let application = await this.applicationRepository.findById(id);
+
+    if (application.services.length === 0) {
+      throw new HttpErrors.UnprocessableEntity(
+        'No service in Application ' + application.id,
+      );
+    }
+
+    let params: {[key: string]: Object} = {
+      application: application,
+    };
+
+    let tenantAssocs = await this.tenantAssociationRepository.find({
+      where: {
+        tenantId: application.tenantId,
+      },
+    });
+
+    if (tenantAssocs.length === 0) {
+      tenantAssocs = await this.tenantAssociationRepository.find({
+        where: {
+          tenantId: 'default',
+        },
+      });
+
+      if (tenantAssocs.length === 0) {
+        throw new HttpErrors.UnprocessableEntity(
+          'No ADC associated with Application ' + application.id,
+        );
+      }
+    }
+
+    let adcs = await this.adcRepository.find({
+      where: {
+        id: tenantAssocs[0].adcId,
+      },
+    });
+
+    if (adcs.length === 0) {
+      throw new HttpErrors.UnprocessableEntity(
+        'Can not find the ADC associated with Application ' + application.id,
+      );
+    }
+    params.adc = adcs[0];
+
+    let serviceId = application.services[0];
+    let service = await this.serviceRepository.findById(serviceId);
+
+    params.service = service;
+
+    if (service.pool) {
+      let pool = await this.poolRepository.findById(service.pool);
+      params.pool = pool;
+
+      if (pool.members.length > 0) {
+        let members = [];
+        for (let member_id of pool.members) {
+          let m = await this.memberRepository.findById(member_id);
+          members.push(m);
+        }
+        params.members = members;
+      } else {
+        params.members = [];
+      }
+    }
+    let rules = [];
+    let wafs = [];
+    if (application.services.length > 0) {
+      let serviceId = application.services[0];
+      let service = await this.serviceRepository.findById(serviceId);
```

Comment:
Some of the logic is duplicate with above. Need to refactor. 
For example: `let service = await this.serviceRepository.findById(serviceId);` was done before.

Reply:
due to the merge. Need to simplify the logic.